[Measurements] Start the workbook

eloiseespel opened the pull request:
#208

tcNickolas commented:
We're doing our workbooks in Jupyter Notebook format - this way they can include Q# code and execute it to highlight some aspects of the solution. Besides, the notebooks (being JSON files) show as plain text in reviews on GitHub, enabling comments, suggestions and all other nice review features. docx files show as binary so any review features are lost.

Would it be possible for you to convert the workbook to notebook format? (I'll do proper content review later, this is just something that jumps at me.)

eloiseespel commented:
Yes I'll work on that right away!

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/QuantumKatas?pullRequest=277) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/QuantumKatas?pullRequest=277) before we can accept your contribution.<br/><br/>:x: eloiseespel [sign now](https://cla.opensource.microsoft.com/microsoft/QuantumKatas?pullRequest=277)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/QuantumKatas?pullRequest=277) it.</sub>

eloiseespel commented:
Sorry about that I had some git struggles, the last commit is the right one. I deleted the doc file and replaced it with the Jupiter Notebook equivalent, I also made the changes you recommended (Thank you so much for the feedback), additionally I added task 1.6 and 1.7. Let me know what you think! I will keep working on the following tasks. 
Thank you for all the feedback, this experience is teaching me a lot!

tcNickolas commented on Measurements/Measurements_Workbook.ipynb:
```diff
@@ -0,0 +1,453 @@
+    "\n",
+    "A workbook is a collection of problems, accompanied by solutions to them. The explanations focus on the logical steps required to solve a problem; they illustrate the concepts that need to be applied to come up with a solution to the problem, explaining the mathematical steps required.\n",
+    "Note that a workbook should not be the primary source of knowledge on the subject matter; it assumes that you've already read a tutorial or a textbook and that you are now seeking to improve your problem-solving skills. You should attempt solving the tasks of the respective kata first and turn to the workbook only if stuck. While a textbook emphasizes knowledge acquisition, a workbook emphas ...
+    "This workbook describes the solutions to the problems offered in the [Measurement Kata](https://notebooks.gesis.org/binder/jupyter/user/microsoft-quantumkatas-iacsymjt/notebooks/Measurements/Measurements.ipynb). Since the tasks are offered as programming problems, the explanations also cover some elements of Q# that might be non-obvious for a first-time user.\n",
```
Please make sure the links are relative, not absolute - the `notebooks.gesis.org/binder/` kind of links leads to a specific instance of virtual environment on Binder that are not visible to anybody but you (and not even to you after the environment shuts down). Relative links will work for everybody, including people who run the katas locally.

```suggestion
    "This workbook describes the solutions to the problems offered in the [Measurement Kata](./Measurements.ipynb). Since the tasks are offered as programming problems, the explanations also cover some elements of Q# that might be non-obvious for a first-time user.\n",
```

tcNickolas commented on Measurements/Measurements_Workbook.ipynb:
```diff
@@ -0,0 +1,453 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "operation IsQubitOne (q : Qubit) : Bool {\n",
```
Please include magic line of the form `%kata T101_IsQubitOne_Test` in the code cells that contain solutions to corresponding tasks - this will allow the user to run those solutions and verify that they are correct while still in the workbook. (The name of the test will change from task to task - you can find it in the corresponding cell of the kata)

tcNickolas commented on Measurements/Measurements_Workbook.ipynb:
```diff
@@ -0,0 +1,453 @@
+   "outputs": [],
+   "source": [
+    "operation InitializeQubit (q : Qubit) : Unit {\n",
+    "\tif(M(q) = One){\n",
```
Please use 4 spaces for indentation instead of tabs - we prefer to have unified style across all Q# code. Also, notebooks highlight tabs, so a mix of tabs and spaces looks weird:
![image](https://user-images.githubusercontent.com/10113024/73402502-4a7c0b80-42a2-11ea-84e2-42734fa89b66.png)

tcNickolas commented on Measurements/Measurements_Workbook.ipynb:
```diff
@@ -0,0 +1,453 @@
+   "metadata": {},
+   "source": [
+    "When a qubit is measured its superposition state collapses, meaning that the superposition state (a linear combination of basis vectors) collapses to just one of the basis vectors. Since in this task the original state of the qubit is already one of the basis vectors, we know measurement would identify that state exactly. \n",
+    "[M( )] (https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.primitive.m?view=qsharp-preview) is the measurement gate, it returns true if the qubit is in state 1 and false if the qubit is in state 0.\n",
```
Markdown doesn't recognize this as link - that extra space breaks the connection between link text and the link itself.
```suggestion
    "[M( )](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.primitive.m?view=qsharp-preview) is the measurement gate, it returns true if the qubit is in state 1 and false if the qubit is in state 0.\n",
```

tcNickolas commented on Measurements/Measurements_Workbook.ipynb:
```diff
@@ -0,0 +1,453 @@
+   "source": [
+    "### Task 1.1.  $|0\\rangle$ or $|1\\rangle$ ?\n",
+    "\n",
+    "**Input:** A qubit which is guaranteed to be in either the $|0\\rangle$ or the $|1\\rangle$ state.\n",
```
Please review the workbook to check that the task description renders in same way as in the original kata. Looks like some line breaks were lost, so input, output and comments are all in the same paragraph:
![image](https://user-images.githubusercontent.com/10113024/73402815-f1f93e00-42a2-11ea-958c-55acc64ae72d.png)
In this task all these sections are small, but later tasks have longer task descriptions, and around task 1.6 they become really hard to read.

tcNickolas commented on Measurements/Measurements_Workbook.ipynb:
```diff
@@ -0,0 +1,453 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "function FindFirstDiff_Reference (bits1 : Bool[], bits2 : Bool[]) : Int {\n",
```
Q# notebooks don't support combining `%kata` magic with multiple functions/operations defined in a single cell, so you'll need to split this cell into two - one with the definition of `FindFirstDiff_Reference` and another one with `%kata` and the definition of the operation `TwoBitstringsMeasurement` itself.

tcNickolas commented on Measurements/Measurements_Workbook.ipynb:
```diff
@@ -0,0 +1,453 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    " operation BasisStateMeasurement_Reference (qs : Qubit[]) : Int {\n",
```
Once you add `%kata` magics, you'll need to name the operations exactly the way they are named in the tasks, otherwise they won't get picked up for testing. This one will have to be `BasisStateMeasurement`

tcNickolas commented on Measurements/Measurements_Workbook.ipynb:
```diff
@@ -0,0 +1,453 @@
+    "}"
+   ]
+  },
+  {
```
Could you please swap solutions with the explanations, so that the explanations come first? We're trying to stick to this order across all workbooks.

eloiseespel commented:
Thank you for all your advice on using git, this is my first time working on open source so I value your expertise, I will clear and re-create the branch. 

And thank you for the comments, I will resolve them asap (I'm actually really enjoying working on this!)

tcNickolas commented:
I'm glad you're enjoying this - it probably won't be the last round of comments I'll leave... :-)

tcNickolas commented:
Any updates on this? Please let me know if you run into any issues with either GitHub or implementing my feedback - I'm happy to help!

tcNickolas commented:
@eloiseespel, are you planning to continue working on this pull request? Looks like it acquired some conflicts over the last month, so now it's definitely easier to copy over your work somewhere, delete this branch and create a new one from the latest master of the main repository. Please let me know if you need more detailed guidance!

eloiseespel commented:
I signed the CLA, thank you for your help cleaning up this file, it is greatly appreciated!
